### PR TITLE
[fix #166] REST endpoint for Subhub to populate Salesforce

### LINKED ISSUE #166: New REST endpoint for Subscription processor to CRM
We will be posting webhook events from payment provider to subscription processor to CRM. Here is a sample request:

```json
{
    "type" : "charge.succeeded",
    "payload" : { "id" : "someid", "someotherfield" : "someothervalue"}
}
```

The payload is generic, the Basket program will have to parse based on the type. The definition for each type is currently being worked on. We are assuming this should affect the signature of the REST endpoint we need from Basket.

### PR DESCRIPTION
**Pull request by jpetto**
Create a simple API to capture Stripe transactions and put them in SFDC/SFMC.

All data will come to a single view/endpoint. That view will check the API token, then route the payload to the appropriate task based on the `event_type` sent in the payload. Tasks perform various bits of data transformation and logic before inserting/updating records in Salesforce.

Fixes #166

### CODE REVIEW

**Comment by jpetto** [comment]
@pmac Still very much in draft mode, but would love a sanity check when you get a chance. 😄

File: basket/news/views.py [hunk 1]
```diff
@@ -939,3 +955,36 @@ def update_user_task(request, api_call_type, data=None, optin=False, sync=False)
...
+    else:
+        etype = data['event_type']
+        processor = SUBHUB_EVENT_TYPES.get(etype)
+
+        if processor:
+            processor.delay(data)
+            return HttpResponseJSON({'status': 'ok'})
+        else:
+            # TODO: set a better HTTP response code? 409?
+            return HttpResponseJSON({'status': 'unknown event type'})
```
Comment (pmac) [inline basket/news/views.py]: At least return a 400 and a status of `error` and this as `desc`. This is good though 👍

File: basket/news/views.py [hunk 2]
```diff
@@ -939,3 +955,36 @@ def update_user_task(request, api_call_type, data=None, optin=False, sync=False)
...
+    if not has_valid_api_key(request):
+        return HttpResponseJSON({
+            'status': 'error',
+            'desc': 'requires a valid API-key',
+            'code': errors.BASKET_AUTH_ERROR,
+        }, 401)
+
+    try:
+        data = json.loads(request.body)
+    except ValueError as _:
```
Comment (pmac) [inline basket/news/views.py]: If you don't need the exception object just leave the `as _` off.
```suggestion
    except ValueError:
```

File: basket/news/views.py [hunk 3]
```diff
@@ -939,3 +955,36 @@ def update_user_task(request, api_call_type, data=None, optin=False, sync=False)
...
+        statsd.incr('subhub_post.message.json_error')
+        sentry_client.captureException(data={'extra': {'request.body': request.body}})
```
Comment (pmac) [inline basket/news/views.py]: 👍

File: basket/news/utils.py [hunk 4]
```diff
@@ -280,9 +280,10 @@ def language_code_is_valid(code):
 ]
 
 
-def get_user_data(token=None, email=None, extra_fields=None, get_fxa=False):
+def get_user_data(token=None, email=None, stripe_id=None, extra_fields=None, get_fxa=False):
```
Comment (pmac) [inline basket/news/utils.py]: Still working on making this work I guess?

File: basket/news/tasks.py [hunk 5]
```diff
@@ -717,6 +717,157 @@ def record_source_url(email, source_url, newsletter_id):
...
+            transaction_data['Credit_Card_Exp__c'] = data['exp_year']  # TODO: check w/marty on SF field names here - they are duplicated in the doc
+
+        try:
+            # will raise a SalesforceMalformedRequest if not found
+            sfdc.opportunity.update('PMT_Transaction_ID__c/{}'.format(charge_id), transaction_data)
+            # TODO: is this the proper syntax? unsure about the use/purpose of the first param here - is there some magic happening that makes this a key?
+        except sfapi.SalesforceMalformedRequest as e:
+            sfdc.opportunity.create(transaction_data)
+    else:
+        # TODO: do we need to do anything here if the user wasn't found?
```
Comment (pmac) [inline basket/news/tasks.py]: Maybe there's enough info in the data to create the user if none is found? Really only need email address.

File: basket/news/utils.py (hunk 4 above)
Comment (jpetto) [reply basket/news/utils.py]: Yeah - just put this in as a starting point.

File: basket/news/tasks.py (hunk 5 above)
Comment (jpetto) [reply basket/news/tasks.py]: According to the doc, the only user identifiable info in these payloads is the Stripe ID, so I don't think we can do much?

File: basket/news/tasks.py [hunk 6]
```diff
@@ -717,6 +717,157 @@ def record_source_url(email, source_url, newsletter_id):
...
+        if event == 'succeeded':
+            transaction_data['Credit_Card_Type__c'] = data['brand']
+            transaction_data['Last_4_Digits__c'] = data['last4']
+            transaction_data['Credit_Card_Exp__c'] = data['exp_month']
+            transaction_data['Credit_Card_Exp__c'] = data['exp_year']  # TODO: check w/marty on SF field names here - they are duplicated in the doc
+
+        try:
+            # will raise a SalesforceMalformedRequest if not found
+            sfdc.opportunity.update('PMT_Transaction_ID__c/{}'.format(charge_id), transaction_data)
+            # TODO: is this the proper syntax? unsure about the use/purpose of the first param here - is there some magic happening that makes this a key?
```
Comment (jpetto) [inline basket/news/tasks.py]: Would love some input on this bit. 😄

Comment (pmac) [reply basket/news/tasks.py]: It's the lookup basically. The first bit is the filed to look in, and the second is the data.

https://github.com/simple-salesforce/simple-salesforce#other-options

**Comment by jpetto** [comment]
@glogiotatidis Do you have time to review this during your morning tomorrow? Tests are only stubs at the moment, but it would be great to get this up on dev for end to end testing of the Salesforce API.

File: basket/news/utils.py [hunk 7]
```diff
@@ -280,9 +280,10 @@ def language_code_is_valid(code):
 ]
 
 
-def get_user_data(token=None, email=None, extra_fields=None, get_fxa=False):
+def get_user_data(token=None, email=None, payee_id=None, extra_fields=None, get_fxa=False):
```
Comment (glogiotatidis) [inline basket/news/utils.py]: This should be fine but generally speaking I'm careful adding more parameters out of order, especially in a function with all optional parameters. I did a quick search and the only place where we don't explicitly name the parameter is in https://github.com/mozmeao/basket/blob/79624a385bcef27c639c312916282b02d11edac6/basket/news/tasks.py#L386, so maybe you can change that to `get_user_data(token=data.get('token'), email=data.get('email'), extra_fields=['id']))` just to be on the safe side.

File: basket/news/tasks.py [hunk 8]
```diff
@@ -717,6 +717,250 @@ def record_source_url(email, source_url, newsletter_id):
...
+    # processor/Stripe ID, so if the user wasn't found by that, there's really
+    # nothing to be done here
+    if user_data:
+        # this is the primary key for the opportunity
+        charge_id = data['charge_id']
+
+        # just in case amount data is missing/malformed
+        try:
+            # amount values are cents - convert to dollars
+            amount_paid = int(data['amount']) / 100
```
Comment (glogiotatidis) [inline basket/news/tasks.py]: This always returns `int`. 
```python
int(108.5) / 100 = 1
```
Just a heads up, not sure if that's what you want here.

File: basket/news/tasks.py [hunk 9]
```diff
@@ -717,6 +717,250 @@ def record_source_url(email, source_url, newsletter_id):
...
+            # set properly in the except block below
+            sfdc.opportunity.update('PMT_Transaction_ID__c/{}'.format(charge_id), transaction_data)
+        except sfapi.SalesforceMalformedRequest:
+            # we are creating a new opportunity, so put the charge_id in
+            transaction_data['PMT_Transaction_ID__c'] = charge_id
+
+            # determine if this is an initial or recurring payment
+            try:
+                # look for existing opportunities with the current subscription_id
+                _ = sfdc.opportunity.get_by_custom_id('PMT_Subscription_ID__C', data['subscription_id'])
```
Comment (glogiotatidis) [inline basket/news/tasks.py]: Why do you need `_`?

File: basket/news/utils.py (hunk 7 above)
Comment (jpetto) [reply basket/news/utils.py]: Ah, nice catch!

I did think maybe `get_user_data` needs a refactor - maybe taking the (now) three identifiers and expecting a dict instead - something like `{ 'type': 'token', 'value': '123456' }`. Anyway, I'll leave that for later clean up.

File: basket/news/tasks.py (hunk 9 above)
Comment (jpetto) [reply basket/news/tasks.py]: Meant to get rid of that - was just coding quickly and wasn't sure if I was going to be using the result. Will remove.

File: basket/news/tasks.py (hunk 8 above)
Comment (jpetto) [reply basket/news/tasks.py]: Hmmm, good catch. Will replace with:

`int(data['amount']) / float(100)`

That should do it, right?

**Comment by jpetto** [comment]
@glogiotatidis Made a commit addressing your feedback. If you have time to verify tomorrow morning, that would be great. Hopefully we can push this out to dev tomorrow. 😁

File: basket/news/tasks.py (hunk 8 above)
Comment (glogiotatidis) [reply basket/news/tasks.py]: Yes that works. Another trick is

`int(data['amount']) / 100.0`

**Comment by glogiotatidis** [comment]
Looks good, thanks!

**Pull request merged**